src/perf_counters.cc

```diff
@@ -0,0 +1,114 @@
+#include "perf_counters.h"
+
+#if defined HAVE_LIBPFM
+#include <cstring>
+
+#include "perfmon/pfmlib.h"
+#include "perfmon/pfmlib_perf_event.h"
+#endif
+
+#ifdef BENCHMARK_OS_WINDOWS
+size_t read(int, void*, size_t) { return 0; }
+#endif
+namespace benchmark {
+namespace internal {
+
+#if defined HAVE_LIBPFM
+const bool PerfCounters::kSupported = true;
+
+bool PerfCounters::Initialize() { return pfm_initialize() == PFM_SUCCESS; }
+
+PerfCounters PerfCounters::Create(
+    const std::vector<std::string>& counter_names) {
+  if (counter_names.empty()) {
+    return NoCounters();
+  }
+  if (counter_names.size() > PerfCounterValues::kMaxCounters) {
+    std::cerr << counter_names.size()
+              << " counters were requested. The minimum is 1, the maximum is "
+              << PerfCounterValues::kMaxCounters << "\n";
+    return NoCounters();
+  }
+  std::vector<int> counter_ids(counter_names.size());
+
+  const int mode = PFM_PLM3;  // user mode only
+  for (size_t i = 0; i < counter_names.size(); ++i) {
+    const bool is_first = i == 0;
+    struct perf_event_attr attr;
+    std::memset(&attr, 0, sizeof(perf_event_attr));
+    attr.size = sizeof(attr);
+    const int group_id = !is_first ? counter_ids[0] : -1;
+    const auto& name = counter_names[i];
+    if (name.empty()) {
+      std::cerr << "A counter name was the empty string\n";
+      return NoCounters();
+    }
+    pfm_perf_encode_arg_t arg;
+    std::memset(&arg, 0, sizeof(pfm_perf_encode_arg_t));
+    arg.attr = &attr;
+
+    const int pfm_get =
+        pfm_get_os_event_encoding(name.c_str(), mode, PFM_OS_PERF_EVENT, &arg);
+    if (pfm_get != PFM_SUCCESS) {
+      std::cerr << "Unknown counter name: " << name << "\n";
+      return NoCounters();
```

Comment:
do you want to skip adding all the others here? we could add the ones we recognise at least and continue (having logged the error). wdyt?

Reply:
it may be lost to the user - failing fast lets them correct and restart.